core/store/src/trie/trie_recording.rs

```diff
@@ -738,3 +786,98 @@ mod trie_recording_tests {
         test_trie_recording_consistency(true, true, true);
     }
 }
+
+#[cfg(test)]
+mod memtrie_batch_iteration_tests {
+    use crate::Trie;
+    use crate::test_utils::{
+        TestTriesBuilder, create_test_store, test_populate_flat_storage, test_populate_trie,
+    };
+    use crate::trie::AccessOptions;
+    use crate::trie::trie_tests::merge_trie_changes;
+    use near_primitives::hash::hash;
+    use near_primitives::shard_layout::ShardUId;
+
+    use super::*;
+
+    /// Returns the hash of height (as le_bytes) for use as a fake block hash in tests.
+    fn fake_hash(height: usize) -> CryptoHash {
+        hash(height.to_le_bytes().as_ref())
+    }
+
+    fn iterate_batch(
+        trie: &Trie,
+        previous_batch_last_key: Option<Vec<u8>>,
+        batch_limit: usize,
+    ) -> Option<Vec<u8>> {
+        let read_trie = trie.lock_for_iter();
+        // Get the iterator for the trie, skipping the first key if needed
+        let mut iter = read_trie.iter().expect("failed to get iterator");
+        if let Some(key) = previous_batch_last_key {
+            iter.seek(&key, true).expect("failed to seek");
+        }
+
+        // Iterate over the trie, stopping when we reach the batch size
+        let mut items = 0;
+        while let Some(result) = iter.next() {
+            let Ok((key, _value)) = result else {
+                panic!("failed to iterate");
+            };
+            if items >= batch_limit {
+                return Some(key);
+            }
+            items += 1;
+        }
+
+        None // No more items to iterate
+    }
+
+    fn test_batched_iteration_impl(use_memtries: bool) {
```

Comment:
Some other test idea - create trie based on recorded storage and check that iterating this trie gives the same result and recorded storage, like in `test_trie_recording_consistency` test.
However, the check is less justified than for state witness, where having the same recorded storage is crucial.

Reply:
That doesn't exactly work because of the difference between PartialStorage and TrieChanges.
But I applied the same idea:
- Create a new store from the TrieChanges,
- Iterate that trie,
- Check we get the same items & recorded TrieChanges